### DISCUSSION TITLE
Loader stl file

### DISCUSSION THREAD

**Discussion by Orphenix** [makimenko/angular-template-for-threejs#375] [Q&A]
Hi, I wanted to know if it is possible to be able to view an example stl file https://ozeki.hu/attachments/116/Menger_sponge_sample.stl.

I have downloaded your project and have performed various tests, but, I have not found any solution to load an STL file, not even on the documentation.
I know, that three.js allows loading https://threejs.org/examples/?q=stl#webgl_loader_stl, can you tell me a way to load a grt file?

Tsk 
Rocco

**Comment by makimenko** [comment]
Hi, 
This is possible in Three.js and that means that it's easy to add into a new component into atft:

Sample of similar component:
https://github.com/makimenko/angular-template-for-threejs/blob/master/projects/atft/src/lib/object/loader/object-loader.component.ts
```js
@Component({
  selector: 'atft-object-loader',
  providers: [provideParent(ObjectLoaderComponent)],
  template: '<ng-content></ng-content>'
})
export class ObjectLoaderComponent extends AbstractModelLoader {
  private loader = new THREE.ObjectLoader();

  constructor(
    protected rendererService: RendererService,
    @SkipSelf() @Optional() protected parent: AbstractObject3D<any>
  ) {
    super(rendererService, parent);
  }

  protected async loadLazyObject() {
    // console.log('ObjectLoaderComponent.loadLazyObject');
    return new Promise<THREE.Object3D>((resolve, reject) => {
      this.loader.load(this.model, model => {
          // BUG #95: it seems that some textures loaded after last render (and model has black texture instead)
          resolve(model);
        },
        undefined,
        reject
      );
    });
  }

}
```


Theoretically, only this should be changed in new component:
```
private loader = new THREE.STLLoader();
```

If you can create PR, then I would be very appreciated.
Otherwise, will I add changes in the next release.

**Reply by Orphenix** [reply]
Thanks I will try and let you know

**Reply by Orphenix** [reply]
Yes I solved it, I did as you said ... but I had to change the method make a change to the result of the STL loader, I had to create a mesh as Three.js suggests. So:

```
export class StlLoaderComponent extends AbstractModelLoader {
  private loader = new STLLoader();

  constructor(
    protected rendererService: RendererService,
    @SkipSelf() @Optional() protected parent: AbstractObject3D<any>
  ) {
    super(rendererService, parent);
  }

  protected async loadLazyObject() {
    // console.log('ObjectLoaderComponent.loadLazyObject');
    return new Promise<THREE.Object3D>((resolve, reject) => {
      this.loader.load(this.model, geometry => {
          const material = new THREE.MeshPhongMaterial();
          const mesh = new THREE.Mesh(geometry, material);
          // BUG #95: it seems that some textures loaded after last render (and model has black texture instead)
          resolve(mesh);
        },
        undefined,
        reject
      );
    });
  }


}

```
se quiser posso abrir uma filial com essa funcionalidade

**Reply by makimenko** [reply]
Created PR: https://github.com/makimenko/angular-template-for-threejs/pull/381
Soon in atft:1.4.21 atft-stl-loader component added:
```xml
    <atft-stl-loader model="yourfile.stl" materialColor="#AAFFAA">
    </atft-stl-loader>
```
![image](https://user-images.githubusercontent.com/11466819/116785909-b56b5200-aaa4-11eb-9356-8c87bff25eac.png)

**Reply by makimenko** [reply]
https://makimenko.github.io/angular-template-for-threejs/storybook/?path=/story/other-loader--stl-loader
